### PR TITLE
ESQL: Check field exists before load from _source

### PR DESCRIPTION
**Pull request by nik9000**
Loading fields from `_source` is *super* slow because you have to decompress the stored fields and then turn the stored field into a map-of-maps. And then dig through the map-of-maps. This adds "does this field exist" style checks before most loads from `_source`. Not all fields can do it, but most fields can.

This really improves the performance of our
`esql_dissect_duration_and_stats` benchmark, mostly because it is running `dissect` on a field that has to load from `_source` that isn't in many of the documents. Here's the performance:
```
|  50th percentile service time | 867.667 | 100.491 | -767.176 | ms | -88.42% |
|  90th percentile service time | 886.042 | 102.434 | -783.608 | ms | -88.44% |
| 100th percentile service time | 893.035 | 104.598 | -788.437 | ms | -88.29% |
```

### CODE REVIEW

**Comment by costin** [comment]
Great stuff - I wonder if this is something we can do at the planner level without affecting the other columns.

File: server/src/main/java/org/elasticsearch/index/mapper/BlockSourceReader.java [hunk 1]
```diff
@@ -262,4 +348,66 @@ static BytesRef toBytesRef(BytesRef scratch, String v) {
...
+     */
+    public static LeafIteratorLookup lookupFromFieldNames(String fieldName) {
+        return new LeafIteratorLookup() {
+            private final BytesRef name = new BytesRef(fieldName);
+
+            @Override
+            public DocIdSetIterator lookup(LeafReaderContext ctx) throws IOException {
+                Terms terms = ctx.reader().terms(FieldNamesFieldMapper.NAME);
+                if (terms == null) {
+                    return null;
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/mapper/BlockSourceReader.java:381]: Should we return `DocIdSetIterator.all(ctx.reader().maxDoc())` when the terms don't exist? I think we allow disabling the `_field_names` field on 7.x. However, we would lose the benefit for fields without values in a segment. We can check the created index version, but we don't have it here.

**Review by dnhatn — COMMENTED** [review]
I left a comment, but this is neat. Thanks Nik!

**Comment by nik9000** [comment]
> Great stuff - I wonder if this is something we can do at the planner level without affecting the other columns.

In some cases I think we can push the `IS NOT NULL` constraint all the way to Lucene. This'll still be helpful in the cases where we can't push it all the way. Like when we actually need the `null`-ness of the field for something.

File: server/src/main/java/org/elasticsearch/index/mapper/BlockSourceReader.java (hunk 1 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/mapper/BlockSourceReader.java:381]: Oh boy! That's fun. Let me play a bit.

**Comment by nik9000** [comment]
@dnhatn I've got it! You were right, the field names field can be disabled, but it's not simple to do. It's kind of annoying to do in a unit test too, so I dropped it into an integration test and then fixed it.

**Review by dnhatn — APPROVED** [review]
LGTM. Thanks Nik for the extra iteration.

**Comment by nik9000** [comment]
Thanks for pointing out the problem @dnhatn ! It'd be lame if indices upgraded from 7.x didn't work properly.

**Pull request merged**